woocommerce/woocommerce
# Call marketing extension recommendations API version 1.3 and support Klaviyo in "Installed extensions" card in Marketing page

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

- Depends on PR https://github.com/Automattic/woocommerce.com/pull/17417.
- Link to P2: pbIy4N-3TH-p2





In this PR, we change the call to marketing extensions recommendations API from version 1.2 to version 1.3 which contains Klaviyo extension, a new recommendation. 

We also make Klaviyo shows up in the "Installed extensions" card in the Marketing page after users install the extension. To do this, I downloaded the Klaviyo extension (latest version 3.1.0) and looked into its code to see how it determines whether setup is completed or not. The way they do it is by checking the existence of `klaviyo_public_api_key` in `klaviyo_settings` WP Option. If `klaviyo_public_api_key` is set, then it means setup is done. We use similar logic to display "Finish setup" button or "Manage" button in the "Installed extensions" card.



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

Pre-requisite: 

- Have the PR https://github.com/Automattic/woocommerce.com/pull/17417 running in your local machine.
- Modify the `get_recommended_plugins` in `MarketingSpecs.php` in the woocommerce plugin to the following, so that the plugin will make the request to the recommendations API version 1.3 running locally in your machine, ignoring the transient cache. You may need to modify the domain, especially if you have changed the port to something else (in my case, I changed it to port 8443).

```php
	/**
	 * Load recommended plugins from WooCommerce.com
	 *
	 * @return array
	 */
	public function get_recommended_plugins(): array {
		$plugins = get_transient( self::RECOMMENDED_PLUGINS_TRANSIENT );

		// if ( false === $plugins ) {
			$request = wp_remote_get(
				'https://woocommerce.test:8443/wp-json/wccom/marketing-tab/1.3/recommendations.json',
				array(
					'user-agent' => 'WooCommerce/' . WC()->version . '; ' . get_bloginfo( 'url' ),
					'sslverify' => false,
				)
			);
			$plugins = [];

			if ( ! is_wp_error( $request ) && 200 === $request['response']['code'] ) {
				$plugins = json_decode( $request['body'], true );
			}

			set_transient(
				self::RECOMMENDED_PLUGINS_TRANSIENT,
				$plugins,
				// Expire transient in 15 minutes if remote get failed.
				// Cache an empty result to avoid repeated failed requests.
				empty( $plugins ) ? 900 : 3 * DAY_IN_SECONDS
			);
		// }

		return array_values( $plugins );
	}
```

- Modify the `case TYPES.SET_RECOMMENDED_PLUGINS` in `plugins/woocommerce-admin/client/marketing/data/reducer.js` to the following. This is to fix the icon path, especially when you have woocommerce.com API running on a different port (in my case, it is port 8443).

```js
		case TYPES.SET_RECOMMENDED_PLUGINS:
			const plugins = action.data.plugins.map( ( p ) => {
				return {
					...p,
					icon: p.icon.replace(
						'https://woocommerce.test',
						'https://woocommerce.test:8443'
					),
				};
			} );
			return {
				...state,
				recommendedPlugins: {
					...state.recommendedPlugins,
					[ action.data.category ]: plugins,
				},
			};
```

Test instructions:

1. In a WooCommerce site, go to WooCommerce > Marketing page: `/wp-admin/admin.php?page=wc-admin&path=%2Fmarketing`.
2. You should see the following page with "Discover more marketing tools" card. Klaviyo should be in the Email tab, between MailChimp and Creative Mail. Klaviyo should have an "Install plugin" button.

![image](https://github.com/woocommerce/woocommerce/assets/417342/b0c5ccec-efa9-4cd7-ad00-d122fc01dec2)

3. Click on the "Install plugin" button. You should see install progress indicator. When installation is finished, Klaviyo should disappear in the "Discover more marketing tools" card, and shows up in the "Installed extensions" card with a "Finish setup" button.

![image](https://github.com/woocommerce/woocommerce/assets/417342/ef061e66-5b79-40eb-9bf8-ef39ca9204e9)

4. Click on the "Finish setup" button, and you should see the following "Connect account" page.

![image](https://github.com/woocommerce/woocommerce/assets/417342/ccfb7832-2ef2-418c-bc78-d369ebba34ff)

5. Proceed to create and connect your Klaviyo account.
6. When the account connection is done, go back to the marketing page. In the "Installed extensions" card, Klaviyo should now have a "Manage" button.

![image](https://github.com/woocommerce/woocommerce/assets/417342/fffef01d-77e2-4d80-aaf5-a77bcb6784f1)

7. Click on the "Manage" button, and you should see the following "Connected" page: 

![image](https://github.com/woocommerce/woocommerce/assets/417342/3fe2e716-8e97-4d17-9a7c-cc43e494983b)



### Changelog entry




<details>

#### Significance




#### Type




#### Message 

Support Klaviyo extension in the "Installed extensions" card in Marketing page.

#### Comment 

</details>
